Sources/SwiftBSON/BSONDocument.swift

```diff
@@ -323,36 +362,44 @@ public struct BSONDocument {
             return self.buffer.writerIndex - writer
         }
 
-        internal func validate() throws {
+        @discardableResult
+        internal func validateAndRetrieveKeys() throws -> Set<String> {
```

Comment:
in order to support passing in the keys to the initializer above, I started returning the keys here.

this also led me to realize that we were not validating here that the BSON didn't contain duplicate keys so I started doing so.